Add redirects/rest api/is_robots into wp-cache-phase2.php

stodorovic opened the pull request:
This PR covers some cases of 'unknown page type':
* Use filter [wp_redirect_status](https://developer.wordpress.org/reference/hooks/wp_redirect_status/) and function [http_response_code](http://php.net/manual/en/function.http-response-code.php) to catch all possible redirects.
* Use filter [status_header](https://developer.wordpress.org/reference/hooks/status_header/)  to catch status code for PHP 5.2 and 5.3.
* Use constants `REST_REQUEST` and `JSON_REQUEST` for REST API. I think that covers all WP versions.
* Use constant `WC_API_REQUEST` for detection of [WooCommerce REST API](http://woocommerce.github.io/woocommerce-rest-api-docs/).
* Use [is_robots](https://developer.wordpress.org/reference/functions/is_robots/) and add it into `$wp_super_cache_query`.

Minor corrections:
* Fix typo for filter `wpsc_only_cache_known_pages`.
* Move filter `wp_cache_ob_callback_filter` after call of `wp_super_cache_query_vars`. It's better because it's possible to use `$wp_super_cache_query`. (eg. for sitemaps/feeds detection)
* Use `$wp_super_cache_query[ 'is_404' ]` instead `$wp_cache_404` (and hook `wp_cache_catch_404` )

stodorovic commented:
There is already `apply_filters` for filter `wp_cache_ob_callback_filter`. It's safe to remove `function_exists( 'apply_filter' ) `. I didn't see that anyone is reported issues related to previous `apply_filters`.

Also, moving  `wp_cache_ob_callback_filter` is helpful. I'll try to handle errors related to text/xml:
_This page contains the following errors:
error on line 2 at column 6: XML declaration allowed only at the start of the document
Below is a rendering of the page up to the first error._

I'm still testing this PR. We need to handle REST API (now they aren't unknown pages) . I didn't add code for this because I'm not sure which is good approach.

stodorovic commented:
I added support for status code 304. WP core sends this status for feeds. (eg. https://github.com/WordPress/WordPress/blob/master/wp-includes/class-wp.php#L454 ). It works only for PHP 5.4 and newer. Together with catching redirects, it detects almost all unknown pages on my servers.

I done couple more optimizations related to order of conditions. I removed additional checking for sitemaps ( because we added this code into `wp_super_cache_query_vars` and I tried to optimize checking). Similar for `empty( $wp_super_cache_query )` - it should be last checking.

donnchawp commented on wp-cache-phase2.php:
```diff
@@ -1156,8 +1179,11 @@ function wp_cache_shutdown_callback() {
 		if ( isset( $wp_super_cache_query[ 'is_feed' ] ) ) {
-			$type = get_query_var('feed');
-			$type = str_replace('/','',$type);
+			if ( $type = get_query_var('feed') ) {
```
Assigning a value inside an if() is something I'd like to avoid just because it makes the code slightly harder to read. Can you move that back out please?

donnchawp commented:
Thanks for reorganising the code. Maybe for this first iteration it should default to not cache "is_rest" requests?

stodorovic commented:
I tried to improve detection of feed type ( https://github.com/stodorovic/wp-super-cache/blob/add-redirects-and-rest-into-query-vars/wp-cache-phase2.php#L1196-L1201 ). I think that's better now. 
Because I added double checking of 404 errors, it's better to use `$wp_super_cache_query[ 'is_404' ]`. I added deprecation for `wp_cache_catch_404`. We can remove this code in the future.
Also, I optimized couple `empty`  in conditions. I think that's more readable now.

stodorovic commented:
I added detection of woocommerce api (constant `WC_API_REQUEST`) and I tried to make code (related to sitemap checking) more readable.

stodorovic commented:
@donnchawp Could we merge this PR? I've already added code which avoid using `get_query_var` in `wp_cache_shutdown_callback`. I've added `$wp_super_cache_query[ 'is_sitemap' ]` which is more flexible for checking is it sitemap.
